update db offline settings (#937)
couchbaselabs/docs-sync-gateway: modules/manage/pages/database-offline.adoc

The state diagram in <<fig-state-diag>> represents the states for Sync Gateway and the connection between it and a Couchbase Server database.

[#fig-state-diag]
.State Change Diagram
image::ROOT:state-diagram-offline-12.png[]

The state diagram illustrates two types of operations:

* *Instance-level operations*: Starting or stopping a Sync Gateway instance affects the connections to all of the databases that the instance serves.
* *Database-level operations*: You can perform operations on specific databases independently.
For example, two databases could be online, while a third database could be taken offline, resynchronized, and then brought back online.

== Sync Gateway Configuration Mode Considerations

The workflow for taking databases offline/online differs depending on your configuration mode:

=== Persistent Configuration

If you use the *Persistent Configuration* mode for Sync Gateway, your configuration is stored in Couchbase Server.

Any changes made to your configuration propagate automatically to all Sync Gateway nodes.
You can take advantage of using a Load Balancer for your configuration changes.
For more information on Load Balancers, see xref:[].

> // Add a link here, please.

Reviewer: Don't forget to fill in this placeholder xref I left for you! lol

Author: I've updated it 😀

== Automatically Taking a Database Offline

Sync Gateway will automatically take a database offline if it loses the database's DCP or TAP feed.
This enables the cause to be investigated and rectified.

To bring the database back online after the cause is addressed and the feed(s) restored, use the appropriate Admin REST API endpoint for your configuration mode:

* *Persistent Configuration*: xref:rest-api:rest_api_admin.adoc#tag/Database-Configuration/operation/post_db-_config[`POST /{db}/_config`] with `{"offline": false}`
* *Non-Persistent Configuration*: xref:rest-api:rest_api_admin.adoc#tag/Database-Management/operation/post_db-_online[`POST /{db}/_online`] (call on each node individually)

== Take a Database Offline Manually

This section describes how to manually take a database offline.

[{tabs}]
====
Persistent Configuration::
+
Use this method when your Sync Gateway configuration is stored in Couchbase Server (recommended).
+
. Make a POST request to xref:rest-api:rest_api_admin.adoc#tag/Database-Configuration/operation/post_db-_config[`/{db}/_config`] with `{"offline": true}`.
+
You can use a Load Balancer for this operation.
+
For example:

> //Insert a full REST API/curl example request here, please!

Reviewer: Did you plan to insert an example or leave it without? Please make sure to clean up the comment/remove the stem if you decide otherwise.

Non-Persistent Configuration::
+
Use this method when your Sync Gateway configuration is stored in local config files (legacy mode).
+
. Make a POST request to xref:rest-api:rest_api_admin.adoc#tag/Database-Management/operation/post_db-_offline[`/{db}/_offline`] on each node individually.
+
Do not use a Load Balancer.
+
For example:

> //Insert a full REST API/curl example request here, please!

Reviewer: Same here, this was a comment that needs to be cleaned up - either add the example or remove :)

====

== Take a Database Online Manually

This section describes how to manually bring a database online.

[{tabs}]
====
Persistent Configuration::
+
Use this method when your Sync Gateway configuration is stored in Couchbase Server (recommended).
+
. Make a POST request to xref:rest-api:rest_api_admin.adoc#tag/Database-Configuration/operation/post_db-_config[`/{db}/_config`] with `{"offline": false}`.
+
You can use a Load Balancer for this operation.
+
For example:

> //Insert a full REST API/curl example request here, please!

Reviewer: Add example or remove please :)

Non-Persistent Configuration::
+
Use this method when your Sync Gateway configuration is stored in local config files (legacy mode).
+
. Make a POST request to xref:rest-api:rest_api_admin.adoc#tag/Database-Management/operation/post_db-_online[`/{db}/_online`] on each node individually.
+
Do not use a Load Balancer.
+
For example:

> //Insert a full REST API/curl example request here, please!

Reviewer: Add example or remove please :)

> 

Reviewer: To support changes I made further down the page, you need to add this attribute: 
```suggestion
:tabs:
```

> == Introduction

Reviewer: ```suggestion
```

Sync Gateway enables a database to be taken offline and brought back online without stopping the Sync Gateway instance and without affecting other databases served by it.

The change of status (online or offline) of a database occurs only in the specific Sync Gateway instance addressed.
It is not reflected by other Sync Gateway instances using that database.

> To affect multiple instances, you need to coordinate individual change operations in each of the required Sync Gateway instances.

Reviewer: ```suggestion
If you want to make changes in multiple Sync Gateway instances, you need to coordinate individual change operations in each instance.
```

> === Use Cases

Reviewer: ```suggestion
== When To Take a Database Offline/Online
```

You may need to take a database offline or bring it back online in the following scenarios:

* Taking a database offline without affecting other databases
* Changing configuration properties for a database (while it is offline) without needing to restart Sync Gateway
* xref:resync.adoc[Resynchronizing a database] while it is offline
* Responding to a lost DCP or TAP feed (Sync Gateway does this automatically)
* Creating a database in an offline state so that the start of service delivery for the database can be postponed or coordinated across Sync Gateway instances

> * Performing a Couchbase Server upgrade

Reviewer: ```suggestion
* You need to take a single database offline without affecting other databases.
* You need to change the configuration properties for a database without restarting Sync Gateway.
* You need to xref:resync.adoc[resynchronize a database] while it is offline.
* You are responding to a lost DCP or TAP feed.
Sync Gateway does this automatically.
* You created a database in an offline state, so that the start of service delivery for the database can be postponed or coordinated across Sync Gateway instances.
* You need to run a Couchbase Server upgrade.
```

=== State Diagram

The state diagram in <<fig-state-diag>> represents the states for Sync Gateway and the connection between it and a Couchbase Server database.

[#fig-state-diag]
.State Change Diagram
image::ROOT:state-diagram-offline-12.png[]

In the state diagram:

> * To the left of the gray dashed line, starting or stopping a Sync Gateway instance affects the connections to all of the databases that the instance serves.

Reviewer: There are multiple dashed lines in the diagram. I'm not sure this is the best way to describe this.

Author: I've updated these changes.

* To the right of the gray dashed line, you perform operations on specific databases.
For example, two databases could be online, while a third database could be taken offline, resynchronized, and then brought back online.

> === Configuration Mode Considerations

Reviewer: ```suggestion
== Sync Gateway Configuration Mode Considerations
```

The workflow for taking databases offline/online differs depending on your configuration mode:

*Persistent Configuration (Recommended)*::
* Configuration is stored in Couchbase Server
* Changes propagate automatically to all Sync Gateway nodes
* DO NOT use `/{db}/_offline` and `/{db}/_online` endpoints (these are node-local only)
* Use `POST /{db}/_config` with `{"offline": true/false}` instead

> * Can use a Load Balancer for configuration changes

Reviewer: ```suggestion
=== Persistent Configuration 

If you use the *Persistent Configuration* mode for Sync Gateway, your configuration is stored in Couchbase Server. 

Any changes made to your configuration propagate automatically to all Sync Gateway nodes. 
You can take advantage of using a Load Balancer for your configuration changes.
For more information on Load Balancers, see xref:[].
// Add a link here, please. 

With this configuration mode, you cannot use the `/{db}/_offline` and `/{db}/_online` endpoints to take your databases offline or bring them back online. 
These endpoints are node-local only.

To take your database online and offline, use `POST /{db}/_config` with `{"offline": true/false}`, instead.
```

*Non-Persistent Configuration (Legacy)*::
* Configuration stored in local config files
* Must coordinate changes across all nodes manually
* Must use `/{db}/_offline` and `/{db}/_online` on each individual node

> * Cannot use a Load Balancer for these operations

Reviewer: ```suggestion
=== Non-Persistent Configuration (Legacy)

If you use the *Non-Persistent Configuration* mode for Sync Gateway, your configuration is stored in local config files. 

You must coordinate any changes to your configuration across all Sync Gateway nodes manually. 
You cannot use a Load Balancer for configuration changes. 

With this configuration mode, you must use the `/{db}/_offline` and `/{db}/_online` endpoints on each individual node you want to take offline or bring online. 
```

== Automatically Taking a Database Offline

Sync Gateway will automatically take a database offline if it loses the database's DCP or TAP feed.
This enables the cause to be investigated and rectified.

To bring the database back online after the cause is addressed and the feed(s) restored, use the appropriate Admin REST API endpoint for your configuration mode:

* *Persistent Configuration*: xref:rest-api:rest_api_admin.adoc#tag/Database-Configuration/operation/post_db-_config[`POST /{db}/_config`] with `{"offline": false}`
* *Non-Persistent Configuration*: xref:rest-api:rest_api_admin.adoc#tag/Database-Management/operation/post_db-_online[`POST /{db}/_online`] (call on each node individually)

== Take a Database Offline Manually

This section describes how to manually take a database offline.

> === Persistent Configuration

Reviewer: Delete this heading. See my larger comment further down. 
```suggestion

```

Use this method when your Sync Gateway configuration is stored in Couchbase Server (recommended).

.Prerequisites

* The database is currently online

.Procedure

. Make a POST request to xref:rest-api:rest_api_admin.adoc#tag/Database-Configuration/operation/post_db-_config[`/{db}/_config`] with `{"offline": true}` (can use Load Balancer)

=== Non-Persistent Configuration

Use this method when your Sync Gateway configuration is stored in local config files (legacy mode).

.Prerequisites

* The database is currently online

.Procedure

> . Make a POST request to xref:rest-api:rest_api_admin.adoc#tag/Database-Management/operation/post_db-_offline[`/{db}/_offline`] on each node individually (do not use Load Balancer)

Reviewer: Ah, okay, now that I see how this is laid out and that the prerequisites are quite simple, I'm going to suggest something better than the headings. 

My suggestion: use tabbed content. 

```suggestion
[{tabs}]
====
Persistent Configuration::
+
Use this method when your Sync Gateway configuration is stored in Couchbase Server (recommended).
+
. Make a POST request to xref:rest-api:rest_api_admin.adoc#tag/Database-Configuration/operation/post_db-_config[`/{db}/_config`] with `{"offline": true}`
+
You can use a Load Balancer for this operation. 
+
For example: 
//Insert a full REST API/curl example request here, please!

Non-Persistent Configuration::
+
Use this method when your Sync Gateway configuration is stored in local config files (legacy mode).
+
. Make a POST request to xref:rest-api:rest_api_admin.adoc#tag/Database-Management/operation/post_db-_offline[`/{db}/_offline`] on each node individually.
+
Do not use a Load Balancer.
+
For example: 
//Insert a full REST API/curl example request here, please!
====
```

Reviewer: Also, I'm pretty sure there are other prerequisites other than just the database being online... Don't you need credentials or some kind of authentication for the REST API? If I'm wrong, sure, fine, but I'm positive there's more the user needs. 

I've omitted it for now in case I'm wrong, but.

== Take a Database Online Manually

This section describes how to manually bring a database online.

=== Persistent Configuration

Use this method when your Sync Gateway configuration is stored in Couchbase Server (recommended).

.Prerequisites

* The database is currently offline

.Procedure

. Make a POST request to xref:rest-api:rest_api_admin.adoc#tag/Database-Configuration/operation/post_db-_config[`/{db}/_config`] with `{"offline": false}` (can use Load Balancer)

=== Non-Persistent Configuration

Use this method when your Sync Gateway configuration is stored in local config files (legacy mode).

.Prerequisites

* The database is currently offline

.Procedure

> . Make a POST request to xref:rest-api:rest_api_admin.adoc#tag/Database-Management/operation/post_db-_online[`/{db}/_online`] on each node individually (do not use Load Balancer)

Reviewer: ```suggestion
[{tabs}]
====
Persistent Configuration::
+
Use this method when your Sync Gateway configuration is stored in Couchbase Server (recommended).
+
. Make a POST request to xref:rest-api:rest_api_admin.adoc#tag/Database-Configuration/operation/post_db-_config[`/{db}/_config`] with `{"offline": false}`
+
You can use a Load Balancer for this operation. 
+
For example: 
//Insert a full REST API/curl example request here, please!

Non-Persistent Configuration::
+
Use this method when your Sync Gateway configuration is stored in local config files (legacy mode).
+
. Make a POST request to xref:rest-api:rest_api_admin.adoc#tag/Database-Management/operation/post_db-_online[`/{db}/_online`] on each node individually.
+
Do not use a Load Balancer.
+
For example: 
//Insert a full REST API/curl example request here, please!
====
```

== Starting with a Database Offline

By default, when Sync Gateway starts, it brings all databases that are defined in the configuration file online.

> To keep a database offline when Sync Gateway starts, you can add the `offline` configuration property to the database configuration properties -- see {rest-api-admin--xref--database-configuration}.

Reviewer: ```suggestion
To keep a database offline when Sync Gateway starts, you can add the `offline` configuration property to the database configuration properties.
For more information, see {rest-api-admin--xref--database-configuration}.
```
